### PR TITLE
Converting Scorcher into Dawnbringer

### PR DESCRIPTION
**Pull request by EnvyDragon**
Reworks existing unit into a more distinct anti-surface/anti-crowd role by turning its attack into a set of wandering beams that deal damage, but tend not to dwell on one target for long. 

Currently fires 1-5 beams (increasing by 1 every 1.5s), dealing 5 damage at their endpoints every 0.2s. 

Known bugs: 
- No warp-in model. This is due to the addition of the GimbalZ actor, which allows it to track targets and aim up and down while moving; without that actor, beam movement tends to look a little odd. Haven't found a workaround yet.

demo_small

### CODE REVIEW

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml [hunk 1]
```diff
@@ -17081,4 +17081,41 @@
...
+        <PeriodicEffectArray value="AP_VoidRayPurifierBeamUseMagazine"/>
+        <PeriodicPeriodArray value="0.1"/>
+    </CEffectCreatePersistent>
+    <CEffectApplyBehavior id="AP_VoidRayPurifierBeamDamageFieldAB">
+        <WhichUnit Value="Target"/>
+        <Behavior value="AP_VoidRayPurifierBeamDamageField"/>
+    </CEffectApplyBehavior>
+    <CEffectEnumArea id="AP_VoidRayPurifierBeamSearch">
+        <AINotifyFlags index="HurtEnemy" value="1"/>
+        <SearchFilters value="Ground;Self,Player,Ally,Neutral,Missile,Stasis,Dead,Hidden,Invulnerable"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:17100]: Doesn't damage allied primary target.  (if you order to attack your stuff or an ally). Check Viking's shredder rounds that have handled that case
Snímek obrazovky pořízený 2024-07-13 19-55-42

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml [hunk 2]
```diff
@@ -3071,4 +3071,45 @@
...
+    </CWeaponLegacy>
+    <CWeaponStrafe id="AP_VoidRayPurifierBeamDummyWeapon">
+        <Effect value="AP_PointDefenseLaserDummy"/>
+        <EditorCategories value="Race:Protoss"/>
+        <Options index="Hidden" value="1"/>
+        <AcquireCallForHelpFilters value="-;Player,Ally,Neutral,Enemy"/>
+        <TargetFilters value="Visible,Ground;Missile,Stasis,Dead,Hidden,Invulnerable"/>
+        <Range value="1"/>
+<!--        loiter radius = "area of effect" for this weapon-->
+        <LoiterRadius value="2"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml:3100]: Shouldn't match it splash area size (0.5 from `AP_VoidRayPurifierBeamSearch` effect)?

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml [hunk 3]
```diff
@@ -3071,4 +3071,45 @@
...
+        <Period value="5"/>
+    </CWeaponStrafe>
+    <CWeaponLegacy id="AP_VoidRayPurifierWeaponInfo">
+        <EditorCategories value="Race:Protoss"/>
+        <Icon value="Assets\Textures\btn-upgrade-protoss-airweaponslevel0.dds"/>
+        <DisplayEffect value="AP_VoidRayPurifierBeamDamage"/>
+        <TargetFilters value="Visible;Player,Ally,Neutral,Enemy"/>
+        <Range value="6"/>
+        <Arc value="360"/>
+        <Period value="0.2"/>
```
Comment (Ziktofel) [inline Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml:3112]: Check the displayed attack speed (for beams it should match Effect's `AP_VoidRayPurifierBeamLaunchPersistent` `PeriodicPeriodArray` value)

**Review by Ziktofel — CHANGES_REQUESTED** [review]

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml (hunk 2 above)
Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml:3100]: Not in this case; the intent is that it's meant to be worse at attacking small individual units, so it's meant to "miss" the target fairly often.

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml (hunk 3 above)
Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/WeaponData.xml:3112]: The `VoidRayPurifierBeamLaunchPersistent` effect determines the cooldown between uses of the "magazine", from my testing; it's a relic of the Carrier Interceptor functionality that I used as a baseline, and it doesn't have much impact unless it has a longer period than the `VoidRayPurifierBeamLaunch` weapon. 

The displayed attack speed should (and currently does) match the period of the `VoidRayPurifierBeamDamageField` behavior on the beams themselves, which is responsible for dealing the damaging effect.

File: Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml (hunk 1 above)
Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:17100]: Viking piercing rounds do something different from what I need, but I have a workaround I can try; I'll just apply a temporary behavior if the target's friendly, then make the search effect look for targets that either have that behavior or are enemies.

Comment (EnvyDragon) [reply Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml]: Fixed, Dawnbringer beams now damage all enemy units, AND friendly targets that their parent is specifically attacking.

**Comment by EnvyDragon** [comment]
Fixed issues, and also did some actor changes that should make the turn-and-target-in-place visual effect a little less janky.

**Comment by Ziktofel** [comment]
Bug: Multiple units don't stack on same target

**Comment by EnvyDragon** [comment]
> Bug: Multiple units don't stack on same target

Fixed, thanks for pointing that out

**Pull request merged**